suite/tests/api/detach_state.c

```diff
@@ -410,6 +410,111 @@ test_thread_func(void (*asm_func)(void))
     sideline_ready_for_detach = false;
 }
 
+/***************************************************************************
+ * Client code
+ */
+
+static void *load_from;
+
+static dr_emit_flags_t
+event_bb(void *drcontext, void *tag, instrlist_t *bb, bool for_trace, bool translating)
+{
+#    ifdef X86
+    /* Test i#5786 where a tool-inserted pc-relative load that doesn't reach is
+     * mangled by DR and the resulting DR-mangling with no translation triggers
+     * DR's (fragile) clean call identication and incorrectly tries to restore
+     * state from (garbage) on the dstack.
+     */
+    instr_t *first = instrlist_first(bb);
+    dr_save_reg(drcontext, bb, first, DR_REG_XAX, SPILL_SLOT_1);
+    instrlist_meta_preinsert(bb, first,
+                             XINST_CREATE_load(drcontext, opnd_create_reg(DR_REG_XAX),
+                                               OPND_CREATE_ABSMEM(load_from, OPSZ_PTR)));
+    dr_restore_reg(drcontext, bb, first, DR_REG_XAX, SPILL_SLOT_1);
+    /* XXX i#4232: If we do not give translations, detach can fail as there are so
+     * many no-translation instrs here a thread can end up never landing on a
+     * full-state-translatable PC.  Thus, we set the translation, and provide a
+     * restore-state event.  i#4232 covers are more automated/convenient way to solve
+     * this.
+     */
+    for (instr_t *inst = instrlist_first(bb); inst != first; inst = instr_get_next(inst))
+        instr_set_translation(inst, instr_get_app_pc(first));
```

Comment:
Since this is a test it's not a concern, just trying to understand what this implies: this means each of these instrs are considered a safe spot? And the instrumentation could be interrupted, and if/when we resume, the instrumentation would be repeated too, right?

Reply:
Yes, we'd restart this whole block.